add functional variant of copy_()

bdhirsh opened the pull request:
In order for functions that use `copy_()` to be able to be properly functionalized, we need an out-of-place `copy()` operator. I added one and made it not-exposed in python.

I could also imagine a better version of this where we actually clean up `copy_()` - ideally, maybe we could make `copy_()` structured (and add an out= variant), and maybe get rid of the `_copy_from()` private op used by external backends by XLA. This seems out of scope for the existing problem with functionalization though.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66101 Test functionalization pass in python
* **#67716 add functional variant of copy_()**
* #67715 bugfix for conditional functionalization

Differential Revision: [D32136008](https://our.internmc.facebook.com/intern/diff/D32136008)

bdhirsh committed:
add functional variant of copy_()

[ghstack-poisoned]

vadimkantorov commented:
Is it equivalent to clone? Or

bdhirsh commented:
> Is it equivalent to clone? Or

@vadimkantorov Using `clone()` actually won't work in this case, for two reasons: `copy_()` take in two tensor arguments, while clone only takes in one. In order to appropriately handle dispatch, both tensor arguments need to participate.

That comes up in the case of using vmap, where a vmap'd function tries to call `batched_tensor.copy_(normal_tensor)`. I need an out-of-place variant of the `copy_()` call here, but I need both tensors to participate in the dispatch (so we see the batched_tensor and dispatch to a vmap kernel).

There's also another reason, which is that `copy_()` has some specific semantics that I don't think get preserved by `clone()`: `a.copy_(b)` preserves the size and dtype of `a` (and some other metadata), while only changing the data of `a`.

bdhirsh committed:
Update on "add functional variant of copy_()"

In order for functions that use `copy_()` to be able to be properly functionalized, we need an out-of-place `copy()` operator. I added one and made it not-exposed in python.

I could also imagine a better version of this where we actually clean up `copy_()` - ideally, maybe we could make `copy_()` structured (and add an out= variant), and maybe get rid of the `_copy_from()` private op used by external backends by XLA. This seems out of scope for the existing problem with functionalization though.




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32136008).

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1229,6 +1229,12 @@
+# This doesn't need to be exposed to python, but is needed for functionalization.
+- func: copy(Tensor self, Tensor src, bool non_blocking=False) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: copy
```
This needs a backward formula, right?

bdhirsh replied:
whoops :)

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1229,6 +1229,12 @@
 
 - func: conv_transpose3d.input(Tensor input, Tensor weight, Tensor? bias=None, int[3] stride=1, int[3] padding=0, int[3] output_padding=0, int groups=1, int[3] dilation=1) -> Tensor
 
+# out-of-place variant of copy_()
+# This doesn't need to be exposed to python, but is needed for functionalization.
```
It would be nice if we could add a test for this somehow... Maybe just write a manual test for torch.ops.aten.copy? Should probably test some forward + backward (so we catch things like setting up the gradient formula incorrectly)

bdhirsh replied:
sounds good - just wrote a gradcheck test for `torch.ops.aten.copy`

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1229,6 +1229,12 @@
 
+# out-of-place variant of copy_()
+# This doesn't need to be exposed to python, but is needed for functionalization.
+- func: copy(Tensor self, Tensor src, bool non_blocking=False) -> Tensor
```
not immediately, since I didn't want to preemptively add a bunch of new out-of-place ops. But I bet that as functionalization is used more broadly then we'll start to have to add more out-of-place ops. Right now, you'll get a TORCH_WARN if you try to functionalize an inplace op with no out-of-place variant.

albanD replied:
Do I understand correctly from our offline conversation that this entry is going to go away?

zou3519 replied:
That was my understanding: that the plan is to insert ".to().expand_as()" instead of having a copy operator

bdhirsh replied:
yup- I'll close this PR and add a new one that special cases the `.to().expand_as()` mapping in the odegen.

bdhirsh committed:
Update on "add functional variant of copy_()"

In order for functions that use `copy_()` to be able to be properly functionalized, we need an out-of-place `copy()` operator. I added one and made it not-exposed in python.

I could also imagine a better version of this where we actually clean up `copy_()` - ideally, maybe we could make `copy_()` structured (and add an out= variant), and maybe get rid of the `_copy_from()` private op used by external backends by XLA. This seems out of scope for the existing problem with functionalization though.


Differential Revision: [D32136008](https://our.internmc.facebook.com/intern/diff/D32136008)

[ghstack-poisoned]

bdhirsh committed:
Update on "add functional variant of copy_()"

In order for functions that use `copy_()` to be able to be properly functionalized, we need an out-of-place `copy()` operator. I added one and made it not-exposed in python.

I could also imagine a better version of this where we actually clean up `copy_()` - ideally, maybe we could make `copy_()` structured (and add an out= variant), and maybe get rid of the `_copy_from()` private op used by external backends by XLA. This seems out of scope for the existing problem with functionalization though.


Differential Revision: [D32136008](https://our.internmc.facebook.com/intern/diff/D32136008)

[ghstack-poisoned]
